File: moviejava/app/src/main/java/com/example/moviejava/ui/base/BaseFragment.java
package com.example.moviejava.ui.base;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.LiveData;
import androidx.viewbinding.ViewBinding;

import com.example.moviejava.model.State;


public abstract class BaseFragment<V extends ViewBinding> extends Fragment {
    private V viewBinding;

    protected V binding() {
        return viewBinding;
    }
    private boolean isCreated = true;

   // public BaseFragment(@LayoutRes int contentLayoutId) {
//        super(contentLayoutId);
//    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        isCreated = true;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (viewBinding == null)
            viewBinding = onHandleViewBinding(inflater, container);
        return viewBinding.getRoot();
    }


    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (isCreated) {
            onHandleObject(viewBinding);
            onHandleEvents();
            isCreated = false;
        }
    }
    protected abstract V onHandleViewBinding(LayoutInflater inflater, ViewGroup container);

    protected abstract void onHandleObject(V rootView);

    protected abstract void onHandleEvents();
    protected <T> void defaultObserve(
            LiveData<State<T>> liveData,
            @Nullable OnErrorListener onErrorListener,
            @Nullable OnFinishListener onFinishListener,
            @Nullable OnLoadingListener onLoadingListener,
            OnSuccessListener<T> onSuccessListener
    ) {
        liveData.observe(this, state -> {
            if (state.isLoading()) {
                if (onLoadingListener != null) {
                    onLoadingListener.onLoading(true);
                }
            }
            if (state.isError()) {
                if (onLoadingListener != null) {
                    onLoadingListener.onLoading(false);
                }
                // handle error base
//                if (state.getThrowable() instanceof HttpError) {
//                    handleHttpError((HttpError) state.getThrowable());
//                }

                if (onErrorListener != null) {
                    onErrorListener.onError(state.getThrowable());
                }
                if (onFinishListener != null) {
                    onFinishListener.onFinish();
                }
            }
            if (state.isSuccess()) {
                if (onLoadingListener != null) {
                    onLoadingListener.onLoading(false);
                }
                onSuccessListener.onSuccess(state.getData());
                if (onFinishListener != null) {
                    onFinishListener.onFinish();
                }
            }
        });
    }

    protected interface OnErrorListener {
        void onError(Throwable throwable);
    }

    protected interface OnFinishListener {
        void onFinish();
    }

    protected interface OnLoadingListener {
        void onLoading(boolean isLoading);
    }

    protected interface OnSuccessListener<T> {
        void onSuccess(T data);
    }

//    protected void handleHttpError(HttpError error) {
//        // Handle HTTP error here
//    }

}

File: movie/app/src/main/java/com/example/movie/ui/base/BaseFragment.kt
package com.example.movie.ui.base

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.LiveData
import androidx.viewbinding.ViewBinding
import com.example.movie.model.State

abstract class BaseFragment<V:ViewBinding> :Fragment() {
    private lateinit var viewBinding: V
    protected val binding get() = viewBinding
    private var isCreated = true
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        isCreated = true
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        if (!::viewBinding.isInitialized) {
            viewBinding = setViewBinding(inflater, container)
        }
        return binding.root
    }
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (isCreated) {
            setupObjects(binding)
            setupEvents()
            isCreated = false
        }
    }

    protected abstract fun setViewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?
    ): V

    protected abstract fun setupObjects(rootView: ViewBinding?)

    protected abstract fun setupEvents()

    protected fun <T> defaultObserver(
        liveData: LiveData<State<T>>,
       // doOnError: ((throwable: Throwable?) -> Unit)? = null,
        doOnFinish: (() -> Unit)? = null,
        doOnLoading: ((Boolean) -> Unit)? = null,
        doOnSuccess: (T?) -> Unit,
    ) {
        liveData.observe(this) { state ->
            if (state.isLoading) {
                doOnLoading?.invoke(true)
            }
            if (state.isError) {
                doOnLoading?.invoke(false)
                // handle error base
//                if (state.throwable is HttpError) {
//                    handleHttpError(state.throwable)
//                }
//
//                doOnError?.invoke(state.throwable)
                doOnFinish?.invoke()
            }
            if (state.isSuccess) {
                doOnLoading?.invoke(false)
                doOnSuccess(state.data)
                doOnFinish?.invoke()
            }
        }
    }
    protected fun <T> defaultObserver(
        liveData: LiveData<T>,
        onSuccess: (T) -> Unit
    ) {
        liveData.observe(this) { t -> onSuccess.invoke(t) }
    }
}